src/browser/speaker.js

```diff
@@ -0,0 +1,109 @@
+"use strict";
+
+/**
+ * Note: Uses AudioContext.createScriptProcessor, which is deprecated,
+ * but which no satisfactory substitute is availble.
+ * @constructor
+ * @param {BusConnector} bus
+ * @suppress {deprecated}
+ */
+function SpeakerAdapter(bus)
+{
+    if(typeof window === "undefined")
+    {
+        return;
+    }
+    if(!window.AudioContext && !window.webkitAudioContext)
+    {
+        console.warn("Web browser doesn't support Web Audio API");
+        return;
+    }
+
+    /** @const @type {BusConnector} */
+    this.bus = bus;
+
+    this.audio_context = new (window.AudioContext || window.webkitAudioContext)();
+
+    this.beep_gain = this.audio_context.createGain();
+    this.beep_gain.gain.value = 0;
+    this.beep_gain.connect(this.audio_context.destination);
+
+    this.beep_oscillator = this.audio_context.createOscillator();
+    this.beep_oscillator.type = 'square';
+    this.beep_oscillator.frequency.value = 440;
+    this.beep_oscillator.connect(this.beep_gain);
+    this.beep_oscillator.start();
+
+    this.beep_playing = false;
+    this.beep_enable = false;
+    this.beep_frequency = 440;
+    this.pit_enabled = false;
+
+    this.dac_processor = this.audio_context.createScriptProcessor(2048, 0, 2);
+    this.dac_processor.onaudioprocess = this.dac_process.bind(this);
+    this.dac_processor.connect(this.audio_context.destination);
+
+    bus.register("pcspeaker-enable", function(yesplease)
+    {
+        this.beep_enable = yesplease;
+        this.beep_update();
+    }, this);
+
+    bus.register("pcspeaker-update", function(data)
+    {
+        var counter_mode = data[0];
+        var counter_reload = data[1];
+        this.pit_enabled = counter_mode == 3;
+        this.beep_frequency = OSCILLATOR_FREQ * 1000 / counter_reload;
+        this.beep_update();
+    }, this);
+
+    if(DEBUG)
+    {
+        this.debug_dac = false;
+        this.debug_dac_out = [];
+        window["speaker_debug_dac_out"] = this.debug_dac_out;
+        window["speaker_debug_start"] = () =>
+        {
+            this.debug_dac = true;
+            setTimeout(() =>
+            {
+                this.debug_dac = false;
+            },250);
+        }
+    }
+}
+
+SpeakerAdapter.prototype.beep_update = function()
+{
+    var current_time = this.audio_context.currentTime;
+
+    if(this.pit_enabled && this.beep_enable)
+    {
+        this.beep_oscillator.frequency.setValueAtTime(this.beep_frequency, current_time);
+        if(!this.beep_playing)
+        {
+            this.beep_gain.gain.setValueAtTime(1, current_time);
+            this.beep_playing = true;
+        }
+    }
+    else if(this.beep_playing)
+    {
+        this.beep_gain.gain.setValueAtTime(0, current_time);
+        this.beep_playing = false;
+    }
+};
+
+SpeakerAdapter.prototype.dac_process = function(event)
+{
+    this.bus.send("speaker-tell-samplerate", this.audio_context.sampleRate);
+    this.bus.send("speaker-process", event);
```

Comment:
Now I think about it, I think the `speaker-process` bus messages needs to be rewritten with web workers in mind. Instead of passing the entire `event` object for the sb16 device to modify, is it better for the sb16 device to message back to the speaker adaptor with the data?

I'm not too sure. If we look at `vga.js` and `screen.js`, `vga.js` seems to be receiving the destination buffer (on which vga.js modifies) via the `screen-tell-buffer` bus message: https://github.com/copy/v86/blob/master/src/vga.js#L258-L261

If web workers were to be used, would modifying the buffer within `vga.js` still get updated in `screen.js`?

Reply:
> If web workers were to be used, would modifying the buffer within vga.js still get updated in screen.js?

Not exactly. If the third argument if omitted, the buffer is copied, so this won't work for us. If the underlying buffer is passed as the third argument (see transfer in [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)), the buffer becomes unusable on the sender, but is transferred without copying.

For vga this works, because you can still use `putImageData` even if the image data buffer has been send to another worker.

I think in this case the same trick as for vga might work. If not, we'll need to send the data from sb16 to the speaker.

But to honest, I don't mind if this code is broken in web workers. I haven't tested the web worker in a long time, so it's probably broken anyway. And as I mentioned on gitter, if I ever try web workers again, I'd probably just put the CPU in the worker and let everything else run in the main thread.

Reply:
Gotcha. I was wondering why there was a third unused argument in `BusConnector.prototype.send`, but now it makes sense.

Yup, I'll try to use the same technique of transferring the buffer ownership in sb16, and see if it works.
I'll still try and make sb16 bus calls compatible with web workers, so if we ever change our minds and try putting other devices in the worker, there's one less obstacle to go through.